### PR TITLE
Draft/aliasing implem

### PR DESCRIPTION
**Pull request by serge-sans-paille**

### CODE REVIEW

File: content/2016-04-18-aliasing-improved.rst [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+
+Where is Identifier Binding Used in Pythran
+===========================================
+
+Identifier binding is used by all Pythran analysis that interact with function
+calls, when they need to know something about the function property, or when
+they want to verify that all the possibles (function) values taken by an
+identifier share the same property. For instance:
+
+    #. Conversion from calls with named arguments to call without named arguments, as in ``zeros(1à, dtype=int)``
```
Comment (pbrunet) [inline content/2016-04-18-aliasing-improved.rst]: Typo?

File: content/2016-04-18-aliasing-improved.rst [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+---------------
+
+A dict is abstracted as an unordered container of its values
+
+.. code:: python
+
+    >>> module = ast.parse('def foo(a, b): return {0: a, 1: b}')
+    >>> result = pm.gather(Aliases, module)
+    >>> Aliases.dump(result, filter=ast.Dict)
+    {0: a, 1: b} => ['|a|', '|b|']
```
Comment (pbrunet) [inline content/2016-04-18-aliasing-improved.rst:212]: No information about key in aliasing? (May be for string?)

File: content/2016-04-18-aliasing-improved.rst [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+---------------
+
+Resulting node alias to the return_alias of called function,
+if the function is already known by Pythran (i.e. it's an Intrinsic)
+or if Pythran already computed it's ``return_alias`` behavior.
+
+.. code:: python
+
+    >>> fun0 = 'def f(a): return a\n'
+    >>> fun1 = 'def foo(b): c = f(b)\n'
```
Comment (pbrunet) [inline content/2016-04-18-aliasing-improved.rst]: The multiline docstring notation is nicer:

``` Python
>>> fun = """def f(a): return a
... def foo(b): c = f(b)"""
```

File: content/2016-04-18-aliasing-improved.rst [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+the returned value and its first argument.
+
+.. code:: python
+
+    >>> result = pm.gather(Aliases, module)
+    >>> Aliases.dump(result, filter=ast.Call)
+    f(b) => ['b']
+
+This also works with intrinsics, e.g ``dict.setdefault`` which
+may create alias between its third argument and the return value
```
Comment (pbrunet) [inline content/2016-04-18-aliasing-improved.rst]: Missing dot at the end?

File: content/2016-04-18-aliasing-improved.rst [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+
+.. code:: python
+
+    >>> fun = 'def foo(a, d): __builtin__.dict.setdefault(d, 0, a)\n'
+    >>> module = ast.parse(fun)
+    >>> result = pm.gather(Aliases, module)
+    >>> Aliases.dump(result, filter=ast.Call)
+    __builtin__.dict.setdefault(d, 0, a) => ['<unbound-value>', 'a']
```
Comment (pbrunet) [inline content/2016-04-18-aliasing-improved.rst:278]: You should not return a "contentOf(d, 0)" as alias? (May be we "should" but Pythran "can't" for now)

File: content/2016-04-18-aliasing-improved.rst [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+----------------
+
+Assignment creates aliasing between lhs and rhs
+
+.. code:: python
+
+    >>> module = ast.parse('def foo(a): c = a ; d = e = c ; {c, d, e}')
+    >>> result = pm.gather(Aliases, module)
+    >>> Aliases.dump(result, filter=ast.Set)
+    {c, d, e} => ['|a|', '|a|', '|a|']
```
Comment (pbrunet) [inline content/2016-04-18-aliasing-improved.rst:399]: Should it be a "set" then?

File: content/2016-04-18-aliasing-improved.rst [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+        return s
+
+Pythran knows that ``s`` is a set of integers :-)
+
+
+.. [0] Except the sanity of the developer, but who never used the ``id`` or ``len`` identifiers?
+
+.. [1] Starting from this note, the identifiers from the `ast <https://docs.python.org/2/library/ast.html>`_ module are used.
+
+.. [git-version] The Pythran commit id used for this article is 0000eeeee
```
Comment (pbrunet) [inline content/2016-04-18-aliasing-improved.rst]: Good id?

File: content/2016-04-18-aliasing-improved.rst (hunk 2 above)
Comment (serge-sans-paille) [inline content/2016-04-18-aliasing-improved.rst:212]: Not yet :-) But there is no reason we could not handle string literals as well,
you're right

File: content/2016-04-18-aliasing-improved.rst (hunk 3 above)
Comment (serge-sans-paille) [inline content/2016-04-18-aliasing-improved.rst]: 100% agree

File: content/2016-04-18-aliasing-improved.rst (hunk 5 above)
Comment (serge-sans-paille) [inline content/2016-04-18-aliasing-improved.rst:278]: Yeh, we should if we supported indexed aliasing for dict, which is not
the case :-/

File: content/2016-04-18-aliasing-improved.rst (hunk 7 above)
Comment (serge-sans-paille) [inline content/2016-04-18-aliasing-improved.rst]: > Good id?

Still changing as I integrate some of your remark on the docstring in
the original code :-)

**Comment by pbrunet** [comment]
Nice post !!

**Comment by serge-sans-paille** [comment]
@pbrunet I did quite a few reformatting commits + taken into account most of your advices

**Pull request merged**